### ISSUE TITLE
Configuration Saving

### ISSUE THREAD

**Issue by Azimer** [Azimer/AziAudio#109]
Configuration does not save.  This needs to be implemented.

**Comment by SwooshyCueb** [comment]
I was actually going to bring this up in my last PR, but I forgot about it.

It seems that most PJ64 plugins use ini files for config saving.

**Comment by Azimer** [comment]
It's something I've been holding back on.  I need to determine how I want to do this.

**Comment by SwooshyCueb** [comment]
Most plugins put their ini files somewhere in the Project64 folder, which honestly I don't think is a fantastic idea since it might cause problems for folks who have PJ64 in their Program Files folder and don't have permissions to modify files there.

As well, I kinda feel like ini files are a relic of the 90's, and I'd like to see us use something more modern, personally, like YAML.

**Comment by cxd4** [comment]
It's a pretty multi-sided issue.  It's why I'd put off remembering the config for my RSP plugin for quite a while--using INI files required msvcrt and I did not want to bother learning how to access the Windows registry.  (I spent enough time trying to clean it.)

The only thing with INI is that it's loosely defined...it's a mostly dead spec by Microsoft with a number of open-to-interpretation issues.  (Of course using the standard GetProfileString blah functions helps avoid having to decide personal interpretations, but INI has become available on other operating systems and is still implemented differently in plain C stdio functions, usually with a .CFG extension or something.)

XML might be more modern (Jabo used it for his texture pack zip loading) and I think more supported by e.g. Microsoft, but I have never heard of YAML before.

I also don't really tend to care about text-based configuration as a general rule of thumb.  They, too, also tend to leave things open to interpretation...like how invalid input should be parsed in a line of text like "SyncGameToAudio=truehahaexceptactuallyfalse".  So I kind of think text-based (like INI) config systems are generally of higher merit when there is no GUI to configure, which in this plugin's case isn't so.  So binary is the most preventative "solution" to questions like that but, also tends to cause you to create the rest of the solution divided somewhere else since most people don't expect their users to use a hex editor all the time just to change settings.

**Comment by Azimer** [comment]
The trouble with registry is not everyone has access to write to registry settings (due to a company security policy).
The trouble with config files is that if we are running from an unwritable media, writing config files causes a fault.

The best solution IMO is to write to the windows profile appdata folder (ie C:\Users\<UserName>\AppData\Local\AziAudio .  Then you can use whatever kind of configuration file format we determine is best since we already have a UI configuration.  If we compile under OSX or Linux, we could easily change the folder location to be a non-windows specific current directory location as a default.

**Comment by cxd4** [comment]
Yeah, I do not think much of the issue on where the config should be kept.  AppData sounds fine to me.  I suppose I could play devil's advocate and say that the current "working directory" is better in that a) the saved configuration is portable between computers (as AppData is unique to one computer) b) AppData storage is relatively hidden from the users.  Some users may not mind this at all but I prefer to know when and where extra space is being used by a program (as well as free up space if it was a huge configuration file), and the most obvious place to look is the file path of the program.

But neither a) nor b) are really more important than being able to remember settings at all so I don't really mind the AppData storage.

You mentioned a possible "fault" when writing config files to unwritable media.  It's been a few years since I ran some emulators off CDs but I have tested my spconfig.exe implementation against "Read Only" files with success.
https://github.com/cxd4/spconfig/blob/master/spconfig.c#L73

**Comment by AmbientMalice** [comment]
I think it is important that for PJ64 Azimer's stores its configuration file a "portable" location that is shared by related cfg files. (Now in fairness, I do think PJ64 would benefit from a PCSX2/Dolphin model where config/save/etc files are stored under Documents.)

As for R/W permissions, PJ64 is *supposed* to have permission to modify the contents of the Config folder. This doesn't seem 100% reliable, though. Frankly, I don't agree with emulators being installed to Program Files because it causes no end of headaches. (Hence my love for emulators that come in compressed archives instead of installers.)

**Comment by cxd4** [comment]
On the other hand, there are not that many configuration settings to remember/re-apply if taking AziAudio.dll to another computer which would not remember the settings.  Therefore it wouldn't be that much trouble to centralize such things to the AppData folder.

Yet on the other, other hand (c wut i did thar?), one of the major drawbacks of using the "main exe folder" instead of AppData, being an error on write to an unwritable disk, commonly would apply only to the exercising of the ability to take the plugins and run them on removable media between different computers in the first place.  Which by itself is already something that can't be done with the AppData approach--at least with the local exe folder approach it could be attempted to take it to another computer with or without success in writing to the file.

So yeah either system is hardly comparable to me; I'm fine with either one.

Fundamentally (in terms of abstract portability) I'm sure it could be possible that even the AppData folder itself or most any other directory on Windows could have been marked read-only so it's more a question of avoidance than prevention.

**Comment by weinerschnitzel** [comment]
Test if the program folder is writable, if it is, keep a copy of the config file with AziAudio.dll. If not, use a platform defined path, for example AppData for Windows. This could be expanded to a user defined path to switch between the AppData folder and the Documents folder, if one so wanted.

**Comment by Talantyyr** [comment]
Please keep the config file in the directory where the plugin is, or at least provide an option to load it from there. I really prefer portable applications as i can easily copy them between devices and settings aren't lost if windows needs to be reinstalled.

**Comment by cxd4** [comment]
Having to re-assign a quieter volume every time I launch a ROM begins to get pretty tedious, too.

So not everybody agrees how this config saving should be implemented--that's no surprise.

While I agree with Talantyyr that keeping the save file visible to the local emulator directory is more simplistic and preferential to Azimer's original idea of using AppData on Windows builds, the latter can still be considered preferential if not just for the sake of write-access security.  I think this configuration is not desirable for the user to really see (especially if it's saved as a binary file, not like an INI or something).  So both methods are perfectly good to me.  We should not argue over this, I think, because whether we do it Azimer's way or the local "EXE folder" storage way is still regardless going to be an improvement over no memorization of config settings whatsoever.

**Comment by cxd4** [comment]
The other reason I don't think AppData vs. "EXE folder" method is that important of a debate, is because most of AziAudio configuration is stuff we never touch.  Probably volume is the only thing.

Therefore who really cares if the AppData system prevents "portable devices" from preserving one or two things we set.

**Comment by Azimer** [comment]
I think for windows at least, I will first check AppData for a file, then check EXE folder, then create a new at AppData, (on file) the create a new in EXE folder.  Finally, it would just use the defaults and not save any configuration.

**Comment by cxd4** [comment]
That order sounds correct to me.

The only disadvantage I think is just extra coding.  It could be simpler algorithm to only do AppData or only do EXE folder, but maintaining both can't be considered risky either.

Then again I think the only advantage might be if someone chooses to cut & paste the config out of AppData into EXE folder.  Otherwise I'm not seeing where checking both folders makes a difference.

**Comment by Azimer** [comment]
The difference is, if the file doesn't exist, it will attempt to write to AppData.  On failure it will write to EXE folder.  So given the instance of someone using this on a flash drive and the AppData folder is unavailable due to security policy, it will default back to the EXE folder which will be on the flashdrive.  So another individiual uses a CD-Rom it will fail both AppData and EXE... so thereby using the defaults.

It is a bunch of dinking around but it is convenient... with the huge caveat the emulator is written in a similar way. :)  It won't be a big deal though.

**Comment by cxd4** [comment]
In that case, yes, it does make a difference.  I was not aware that AppData writes could fail.

* If it's on a CD-ROM, we should check AppData first (which is writable), then the CD-ROM.
* If it's on a flash drive, maybe check EXE folder first (which is writable), if AppData is read-only?
* If both the EXE and AppData folders are read && write permissions, prefer AppData.

IDK.  Both the CD-ROM && security policy things sound unusual to the point where I'd just pick a single method to care about instead of using both, but if it's easy to implement, it's easy to implement.

**Comment by Azimer** [comment]
Agreed... but it is easy to implement.

**Comment by Azimer** [comment]
Notes:  cxd4's rudimentary saving was implemented a long while ago and was expanded for the last commit.  A better more robust save needs to be implemented for the final version.

**Comment by cxd4** [comment]
Better and more robust in what way?  Like get it to be text-based?

I would never have done a binary format config in my RSP plugin if I wasn't just seriously trying to remove msvcrt.dll from the dependency list.  For a larger plugin compiled by MSVC such as this I think that becomes less significant; I'd like to sit down and look into a text-based config format.  Not necessarily INI although that's close enough to the traditional Unix config text files on my disk right now so whatever.
